### PR TITLE
Touchscreen interface and driver

### PR DESCRIPTION
**Pull request by tixy**
In implementing support for the touchscreen on the MPS2 board I needed an interface to write the driver to, so I first came up with a simple generic touchscreen API. I don't know if this is suitable or if I should have just made this device specific. Likewise the functions for translating touchscreen values to screen coordinates are generically useful, so I made them so.

### CODE REVIEW

File: include/drivers/touchscreen.h [hunk 1]
```diff
@@ -87,4 +87,63 @@ static inline void touchscreen_set_callback(struct device *dev,
...
+	s64_t	D;
+	s64_t	E;
+	s64_t	F;
+};
+
+/**
+ * @brief Initialise touchscreen coordinate translation parameters.
+ *
+ * This requires that three different and valid display/touchscreen coordinate
+ * pairs are know.
```
Comment (pfalcon) [inline include/drivers/touchscreen.h]: Typo: "known"

File: include/drivers/touchscreen.h [hunk 2]
```diff
@@ -87,4 +87,63 @@ static inline void touchscreen_set_callback(struct device *dev,
...
+ *
+ * @param xlat		Translation parameters to be initialised.
+ * @param display	Three display coordinates that correspond to the given
+ *			touchscreen coordinates.
+ * @param touchscreen	Three touchscreen coordinates that correspond to the
+ *			given display coordinates.
+ */
+void touchscreen_set_calibration(struct touchscreen_xlat *xlat,
```
Comment (pfalcon) [inline include/drivers/touchscreen.h]: If xlat is "Translation parameters to be initialised" (output), is the argument order right here?

File: include/drivers/touchscreen.h [hunk 3]
```diff
@@ -87,4 +87,63 @@ static inline void touchscreen_set_callback(struct device *dev,
...
+ * Convert a touchscreen X,Y sample into the corresponding display X,Y
+ * coordinates by using the previously initialised translation parameters.
+ *
+ * @param xlat		Translation parameters.
+ * @param dislpay_x	Display X coordinate
+ * @param dislpay_x	Display Y coordinate
+ * @param touch_x	Touchscreen X sample
+ * @param touch_y	Touchscreen Y sample
+ */
+void touchscreen_translate(struct touchscreen_xlat *xlat,
```
Comment (pfalcon) [inline include/drivers/touchscreen.h]: And here the order of arguments is: input, output, output, input, input. Is this natural and gives a user the least surprise?

File: drivers/touchscreen/touchscreen.c [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+ * Initialise translation constants using 3 touchscreen coordinates which
+ * correspond to known display coordinates. These constants will be used by
+ * subsequent calls to touchscreen_translate to translate arbitrary points.
+ */
+void touchscreen_set_calibration(struct touchscreen_xlat *xlat,
+				 const struct touchscreen_point display[3],
+				 const struct touchscreen_point touchscreen[3])
+{
+	/* Use shorter aliases for coords */
+	const struct touchscreen_point *T = touchscreen;
```
Comment (pfalcon) [inline drivers/touchscreen/touchscreen.c]: I'm not sure if maintainers would agree with usage of all capitals for something else but macros (variable name in this case).

**Review by pfalcon — COMMENTED** [review]
This looks interesting and useful, I wonder if it would warrant sending an RFC to mailing list, trying to catch more attention with this introduction of new subsystem...

I eyeballed just "touchscreen: Add library for translating coordinates" commit, usual concerns whether teh API signatures are the most natural ones.

**Comment by tixy** [comment]
On Mon, 2017-05-08 at 08:23 -0700, Paul Sokolovsky wrote:
> +				 const struct touchscreen_point display[3],
> 
> +				 const struct touchscreen_point touchscreen[3]);
> +
> +/**
> + * @brief Translate touchscreen coordinates into screen coordinates.
> + *
> + * Convert a touchscreen X,Y sample into the corresponding display X,Y
> + * coordinates by using the previously initialised translation parameters.
> + *
> + * @param xlat		Translation parameters.
> + * @param dislpay_x	Display X coordinate
> + * @param dislpay_x	Display Y coordinate
> + * @param touch_x	Touchscreen X sample
> + * @param touch_y	Touchscreen Y sample
> + */
> +void touchscreen_translate(struct touchscreen_xlat *xlat,
> 
> And here the order of arguments is: input, output, output, input, input. Is this natural and gives a user the least surprise?

I don't know, it seemed natural to me. Guess in my head, the first
parameter is the 'this' pointer if it was C++, i.e. it's the object
being initialise by touchscreen_set_calibration() and the state used by
touchscreen_translate().

**Pull request closed without merging**